cf-file-scanner/src/main/java/org/broadinstitute/ddp/cf/FileScanner.java

```diff
@@ -0,0 +1,280 @@
+
+public class FileScanner implements BackgroundFunction<FileScanner.Message> {
+
+    private static final Logger logger = Logger.getLogger(FileScanner.class.getName());
+
+    private static final String ENV_GCP_PROJECT = "GCP_PROJECT";
+    private static final String ENV_RESULT_TOPIC = "RESULT_TOPIC";
+    private static final String ENV_REFRESH_TIME = "REFRESH_TIME";
+    private static final String ENV_REFRESH_UNIT = "REFRESH_UNIT";
+
+    private static final String ATTR_BUCKET_ID = "bucketId";
+    private static final String ATTR_OBJECT_ID = "objectId";
+    private static final String ATTR_SCAN_RESULT = "scanResult";
+
+    private static final String FRESHCLAM_CONF = "./freshclam.conf";
+    private static final String FRESHCLAM_BIN = "./clamav/freshclam";
+    private static final String CLAMSCAN_BIN = "./clamav/clamscan";
+    private static final String BIN_DIR = "./clamav";
+    private static final String DB_DIR = "/tmp/clamav-db";
+
+    private static final long refreshMillis;
+    private static final Storage storage;
+    private static final Publisher publisher;
+    private static Instant dbUpdatedAt = null;
+
+    // This is ran once on cold start.
+    static {
+        TimeUnit unit = TimeUnit.valueOf(getEnvOrThrow(ENV_REFRESH_UNIT));
+        if (unit == TimeUnit.MICROSECONDS || unit == TimeUnit.NANOSECONDS) {
+            throw new RuntimeException("Refresh time unit is too granular: " + unit);
+        }
+        long refreshTime = Long.parseLong(getEnvOrThrow(ENV_REFRESH_TIME));
+        refreshMillis = unit.toMillis(refreshTime);
+
+        String gcpProjectId = getEnvOrThrow(ENV_GCP_PROJECT);
+        String resultTopic = getEnvOrThrow(ENV_RESULT_TOPIC);
+        try {
+            storage = StorageOptions.newBuilder()
+                    .setCredentials(GoogleCredentials.getApplicationDefault())
+                    .setProjectId(gcpProjectId)
+                    .build().getService();
+        } catch (IOException e) {
+            throw new RuntimeException("Error initializing storage service", e);
+        }
+
+        try {
+            var topic = ProjectTopicName.of(gcpProjectId, resultTopic);
+            publisher = Publisher.newBuilder(topic).build();
+        } catch (IOException e) {
+            throw new RuntimeException("Error initializing pubsub publisher", e);
+        }
+
+        // Pre-load the database files.
+        checkDatabaseFiles();
+    }
+
+    private static String getEnvOrThrow(String name) {
+        String value = System.getenv(name);
+        if (value == null || value.isBlank()) {
+            throw new RuntimeException("Missing environment variable: " + name);
+        }
+        return value;
+    }
+
+    // This method needs to be thread-safe since it uses shared global state and is used in `accept()`.
+    private static synchronized void checkDatabaseFiles() {
+        if (dbUpdatedAt == null || Instant.now().toEpochMilli() - dbUpdatedAt.toEpochMilli() >= refreshMillis) {
```

Comment:
Don't know much about cloud functions. Does this end up being executed by GCP as a single long-lived process? Trying to figure if dbUpdatedAt is truly a global for the whole system.

Reply:
I'm not too experienced with Cloud Function infrastructure either. But what I know now is:
* GCP will create an "instance" to host your function (I think it's app engine under the hood, not sure)
* Global stuff is ran once. Imagine if GCP called `var fn = new YourFunctionClass()` once.
* The `accept()` method is invoked for each request. E.g. imagine GCP calling `fn.accept()`.
* The same function instance MAY be reused for multiple requests, OR new function instances might be spinned up.

So answer to your question is not really. It's only lives long enough as GCP decides it should.

Reply:
So is it big deal that we don't persist the timestamp for last check. Obvious consequence is extra checks and extra downloads.

Reply:
It's not a big deal. It'll just do the download before running a file scan. I added this as a small optimization in case the function instance does get reused (which most likely it will during higher request rates) -- it shaves off roughly 40 seconds.

Reply:
I think it might be best to keep it as-is (with timestamp) if it allows you to fetch the most recent update. Afaik Google keeps these instances sometimes running for a day or so, so the cached version might actually go out of date. But if you'd like to remove this check, no big deal! Then I'd recommend to just cache it once at the start of the function, and reuse on all subsequent invocations.